### ISSUE TITLE
PubSub messages not acknowledged after calling subscriber `.stop!` even when allowed time to finish processing

### ISSUE THREAD

**Issue by joerivett** [googleapis/google-cloud-ruby#11146]
Firstly thank you for a great library and apologies if I'm misinterpreting the docs in this issue report.

The [PubSub Subscriber docs](https://googleapis.dev/ruby/google-cloud-pubsub/latest/Google/Cloud/PubSub/Subscriber.html#stop!-instance_method) say
> #stop!(timeout = nil): Stop this subscriber and block until the subscriber is fully stopped and all received messages have been processed or released, or until timeout seconds have passed.

My interpretation of this is that passing in `timeout` gives currently processing messages time to complete and be acknowledged, however this doesn't appear to be the case. When we call stop! we are seeing the messages complete their callbacks and being acknowledged via `received_message.acknowledge!`, but the acknowledgement doesn't seem to work and the messages are redelivered to another subscriber. Is this expected behaviour?

#### Environment details

  - OS: Ubuntu 16.04.6
  - Ruby version: 2.6
  - Gem name and version: google-cloud-pubsub 2.3.1

#### Steps to reproduce

  1. Create a subscriber with a simple long-running callback e.g. code example below
  2. Publish message to topic
  3. Observe message delivery and processing begin
  4. Stop the subscriber via an interrupt, e.g. Ctrl+C
  5. Observe the message acknowledgement output
  6. Restart subscriber. Observe same message being delivered

#### Code example

```ruby
require "google/cloud/pubsub"

pubsub = Google::Cloud::PubSub.new(
  project_id: "something"
)

# Retrieve a subscription
sub = pubsub.subscription "joe-test-sub"

subscriber = sub.listen do |message|
  puts "Received #{message.message_id}"
  sleep 10
  message.acknowledge!
  puts "Acked #{message.message_id}"
end

# Handle exceptions from listener
subscriber.on_error do |exception|
  puts "Exception: #{exception.class} #{exception.message}"
end

at_exit do
  puts "Stopping subscriber"
  subscriber.stop!(20)
end

subscriber.start

sleep
```

**Comment by quartzmo** [comment]
@joerivett Thanks for opening this issue! I haven't had time yet to run your example. Can you post the output you see when you run it?

### Regarding the documentation for `Subscription#stop!`

The `timeout` arg is ultimately passed to [`Concurrent::ThreadPoolExecutor#wait_for_termination`](https://ruby-concurrency.github.io/concurrent-ruby/1.1.4/Concurrent/ThreadPoolExecutor.html#wait_for_termination-instance_method), which has the following docs:

> Block until executor shutdown is complete or until `timeout` seconds have passed.

That's pretty similar to the `#stop!` description, but does it clear up the situation? Here's a bit more from the `ThreadPoolExecutor` docs:

> There is no way for the thread pool to know what resources are being used by the in-progress tasks. When those tasks are killed the impact on those resources cannot be predicted... Call `#wait_for_termination` with an appropriate `timeout` interval an allow the orderly shutdown to complete

It appears that the timeout value of `20` should be enough for **one** message to be processed given the `sleep 10` statement in the block.

### Regarding Subscription.ack_deadline_seconds 

What does`Subscription#deadline` return for this subscription? The API docs for [`Subscription.ack_deadline_seconds`](https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#google.pubsub.v1.Subscription) state that:

> If this parameter is 0, a default value of 10 seconds is used.

It looks like the `sleep 10` might be just enough for your acknowledgement to miss the deadline.

I don't see the default value of 10 seconds mentioned in this library's docs for [`Topic#subscribe`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/topic.rb#L341-L343). It should probably be added.

**Comment by joerivett** [comment]
Hi @quartzo! Thanks for your speedy response.
I should have clarified about ack deadlines in the initial issue - apologies for missing that (vital!) information. If the ack deadline is 10 seconds, then it makes sense for the message to be redelivered since it may just miss that deadline as you say. However in cases where the deadline is longer, let's say 20 seconds, the message appears to be redelivered as soon as this expires, such as:
1. Start subscriber, `puts sub.deadline => 20`
2. Observe message delivery via "Received {message_id}" output
3. Immediately kill subscriber
4. Wait ~10 seconds to end of sleep, see "Acked {message_id}" output, and program exits
5. Immediately restart subscriber
6. Wait ~10 seconds, see "Received {message_id}" output again
This implies even though the message is given 20 seconds to stop processing (via `stop!(20)`) and the acknowledgement is coming within the deadline, the message still gets redelivered. With that in mind I've slightly modified the above example code to include times in the outputs:

```ruby
require "google/cloud/pubsub"

pubsub = Google::Cloud::PubSub.new(
  project_id: "something"
)

# Retrieve a subscription
sub = pubsub.subscription "joe-test-sub"

puts "Acknowledgement deadline: #{sub.deadline}"

subscriber = sub.listen do |message|
  puts "Received #{message.message_id} at #{Time.now.strftime('%H:%M:%S')}"
  sleep 10
  message.acknowledge!
  puts "Acked #{message.message_id} at #{Time.now.strftime('%H:%M:%S')}"
end

at_exit do
  puts "Stopping subscriber at #{Time.now.strftime('%H:%M:%S')}"
  subscriber.stop!(20)
end

subscriber.start

sleep
```

The output of this is:
```
$ ruby example.rb
Acknowledgement deadline: 20
Received 2292748881611240 at 09:02:30
^CStopping subscriber at 09:02:32
Acked 2292748881611240 at 09:02:40
Traceback (most recent call last):
	1: from example.rb:26:in `<main>'
example.rb:26:in `sleep': Interrupt

$ ruby example.rb
Acknowledgement deadline: 20
Received 2292748881611240 at 09:02:53
Acked 2292748881611240 at 09:03:03
```

Hopefully this helps clarify what I'm seeing. I've dived a little further into the google-cloud-pubsub code - I believe the issue is that the `TimedUnaryBuffer` which actually queues up acknowledgement requests and flushes these every 1 second is stopped immediately when subscriber.stop!() is called. This means any messages that are queued after `stop!()` (i.e. any messages that are permitted to continue processing within the stop timeout, and then get subsequently acknowledged) never have their acknowledgement request sent from the client. I'm unsure if this is by design but hope that is useful knowledge here!

**Comment by quartzmo** [comment]
@joerivett Thanks for digging into this! I'll try to reproduce what you have done soon.

**Comment by quartzmo** [comment]
I reproduced the example above with the following output:

```
quartzmo@Chriss-iMac google-cloud-pubsub % ruby example.rb 
Acknowledgement deadline: 20
Received 2291307010183236 at 10:59:46
^CStopping subscriber at 10:59:53
Acked 2291307010183236 at 10:59:56
Traceback (most recent call last):
	1: from example.rb:24:in `<main>'
example.rb:24:in `sleep': Interrupt

quartzmo@Chriss-iMac google-cloud-pubsub % ruby example.rb
Acknowledgement deadline: 20
Received 2291307010183236 at 11:00:13
Acked 2291307010183236 at 11:00:23
```

**Comment by quartzmo** [comment]
Changing this issue to `type: bug`. 

The documentation for [`Subscriber#stop!`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb#L181-L183) states:

> Stop this subscriber and block until the subscriber is fully stopped and all received messages have been processed or released, or until `timeout` seconds have passed.

But the behavior we are seeing is that the acknowledgement request that is queued in the `TimedUnaryBuffer` is not being sent to the service, thus the received message has not been processed.

**Comment by quartzmo** [comment]
This was fixed in [google-cloud-pubsub v2.6.1](https://rubygems.org/gems/google-cloud-pubsub/versions/2.6.1), thanks!

**Comment by joerivett** [comment]
@quartzmo thanks so much!
